lib/native/http.ts

```diff
@@ -115,76 +116,70 @@ export class HttpClientConnection extends HttpConnection {
         return stream = new HttpClientStream(
             native_handle,
             this,
-            request,
-            on_response,
-            on_body);
+            request);
     }
 }
 
-class HttpStream extends NativeResource implements ResourceSafe {
-    public readonly complete: Promise<Number>;
-    private resolve_complete?: (error_code: Number) => void;
-    private reject_complete?: (reason: any) => void;
-
+class HttpStream extends NativeResourceMixin(BufferedEventEmitter) implements ResourceSafe {
     protected constructor(
         native_handle: any,
-        public connection: HttpConnection,
-        protected on_body_cb?: StreamBodyCallback) {
-        super(native_handle);
-        this.complete = new Promise((resolve, reject) => {
-            this.resolve_complete = resolve;
-            this.reject_complete = reject;
-        });
+        public connection: HttpConnection) {
+        super();
+        this._super(native_handle);
+        this.cork();
     }
 
     close() {
         crt_native.http_stream_close(this.native_handle());
     }
 
-    _on_body(data: ArrayBuffer) {
-        if (this.on_body_cb) {
-            this.on_body_cb(data);
+    on(event: 'end', listener: () => void): this;
+    on(event: 'response', listener: (status_code: number, headers: HttpHeaders) => void): this;
+    on(event: 'headers', listener: (headers: HttpHeaders) => void): this;
+    on(event: 'data', listener: (body_data: ArrayBuffer) => void): this;
+    on(event: 'error', listener: (error: Error) => void): this;
+    on(event: 'ready', listener: () => void): this;
```

Comment:
what's `ready`?

Reply:
Every stream has this 'ready' event, it's the standard way to say "the stream is going to start streaming now"